Lib/asyncio/streams.py

```diff
@@ -154,26 +154,22 @@ def connection_lost(self, exc):
         # Wake up the writer if currently paused.
         if not self._paused:
             return
-        waiter = self._drain_waiter
-        if waiter is None:
-            return
-        self._drain_waiter = None
-        if waiter.done():
-            return
-        if exc is None:
-            waiter.set_result(None)
-        else:
-            waiter.set_exception(exc)
+
+        while self._drain_waiters:
+            waiter = self._drain_waiters.popleft()
+            if not waiter.done():
+                if exc is None:
+                    waiter.set_result(None)
+                else:
+                    waiter.set_exception(exc)
 
     async def _drain_helper(self):
         if self._connection_lost:
             raise ConnectionResetError('Connection lost')
         if not self._paused:
             return
-        waiter = self._drain_waiter
-        assert waiter is None or waiter.cancelled()
         waiter = self._loop.create_future()
-        self._drain_waiter = waiter
+        self._drain_waiters.append(waiter)
         await waiter
```

Comment:
I think it would be better to follow the pattern in Event here so the waiter is removed from the deque when it's cancelled:

https://github.com/python/cpython/blob/16b6e14cc291865ecaf50fa57d0685da94496567/Lib/asyncio/locks.py#L209-L215

```suggestion
        self._drain_waiters.append(waiter)
        try:
            await waiter
        finally:
            self._drain_waiters.remove(waiter)
```